mrmustard/physics/batches.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains the Batch class.
+"""
+
+# pylint: disable=too-many-instance-attributes
+
+from __future__ import annotations
+from typing import Any, Collection, Iterable
+
+import string
+import random
+
+from mrmustard import math
+from mrmustard.utils.typing import (
+    ComplexMatrix,
+    ComplexTensor,
+    ComplexVector,
+    Scalar,
+)
+
+__all__ = ["Batch"]
+
+
+class Batch:
+    r"""
+    The class responsible for keeping track of and handling batch dimensions.
+
+    Args:
+        data: The batched array.
+        batch_shape: The (optional) shape of the batch dims. Defaults to the first dimension of ``data``.
+        batch_labels: The (optional) labels for the batch dims. Defaults to random characters.
+    """
+
+    def __init__(
+        self,
+        data: ComplexMatrix | ComplexVector | ComplexTensor,
+        batch_shape: tuple[int, ...] | None = None,
+        batch_labels: tuple[str, ...] | None = None,
+    ):
+        self._data = data
+        self.dtype = self._data.dtype
+        self._batch_shape = batch_shape or self._data.shape[:1]
+        if self._data.shape[: len(self._batch_shape)] != self._batch_shape:
+            raise ValueError(
+                f"Invalid batch shape {self._batch_shape} for data shape {self._data.shape}."
+            )
+        self._batch_labels = (
+            batch_labels
+            if batch_labels
+            else tuple((random.choice(string.ascii_letters) for _ in self._batch_shape))
+        )
+        self._core_shape = self._data.shape[len(self._batch_shape) :]
+
+    @property
+    def batch_labels(self) -> tuple[str, ...]:
+        r"""
+        The batch labels.
+        """
+        return self._batch_labels
+
+    @property
+    def batch_shape(self) -> tuple[int, ...]:
+        r"""
+        The batch shape.
+        """
+        return self._batch_shape
+
+    @property
+    def core_shape(self) -> tuple[int, ...]:
+        r"""
+        The core shape.
+        """
+        return self._core_shape
+
+    @property
+    def data(self) -> ComplexMatrix | ComplexVector | ComplexTensor:
+        r"""
+        The underlying batched data.
+        """
+        return self._data
```

Comment:
why not just attributes? (since there's no computation nor lazy evaluation?)

Reply:
This is a great question. The main difference between using property and attribute is that attribute comes with a getter, setter and delete whereas you need to explicitly define those for property. Generally, I prefer the latter for this exact reason as more explicit definitions of these methods is preferred (e.g. including validation on the setter). In this context however, I want to keep it as a property as I want the `Batch` class to be immutable. Being able to set the data we would need to also update core_shape, batch_shape and batch_labels appropriately and I think it's just easier to lock it down.